atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/transaction/impl/SnapshotTransaction.java

```diff
@@ -1174,8 +1163,25 @@ protected <K extends Exception> void batchAcceptSizeHint(
 
     private void validatePreCommitRequirementsOnReadIfNecessary(
             TableReference tableRef, long timestamp, boolean allPossibleCellsReadAndNonEmpty) {
+        UUID readId = UUID.randomUUID();
+        if (!allPossibleCellsReadAndNonEmpty) {
+            readsRequiringImmutableTimestampValidationAtCommitRound.add(readId);
+        }
+
         if (isValidationNecessaryOnReads(tableRef, allPossibleCellsReadAndNonEmpty)) {
             throwIfPreCommitRequirementsNotMet(null, timestamp);
+
+            // It may not be present, but that's OK.
+            readsRequiringImmutableTimestampValidationAtCommitRound.remove(readId);
+        }
+    }
+
+    private void validatePreCommitRequirementsOnNonExhaustiveReadIfNecessary(TableReference tableRef, long timestamp) {
+        UUID readId = UUID.randomUUID();
+        readsRequiringImmutableTimestampValidationAtCommitRound.add(readId);
+        if (isValidationNecessaryOnReads(tableRef, false)) {
+            throwIfPreCommitRequirementsNotMet(null, timestamp);
+            readsRequiringImmutableTimestampValidationAtCommitRound.remove(readId);
         }
```

Comment:
This makes sense, though probably needs a few comments - in particular, that we track this because we can't guarantee the task is single-threaded when the transaction attempts to commit (if we could assume that, then this method is much easier written as `(if validation necessary then validate else add a UUID to readsRequiring)`).